### PR TITLE
rnn running on Spark

### PR DESCRIPTION
**Pull request by JerryYanWan**
## What changes were proposed in this pull request?

Modify rnn to make it accessable to Spark.

## How was this patch tested?
To see it can properly run on Spark local.
The loss converges.

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/SimpleRNN.scala [hunk 1]
```diff
@@ -26,10 +26,9 @@ object SimpleRNN {
   outputSize: Int)
   : Module[Float] = {
     val model = Sequential[Float]()
-      .add(Recurrent[Float](hiddenSize)
-        .add(RnnCell[Float](inputSize, hiddenSize, Tanh[Float]())))
-      .add(Select(1, 1))
-      .add(Linear[Float](hiddenSize, outputSize))
+    model.add(Recurrent[Float](hiddenSize)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/SimpleRNN.scala]: Shall we remove hiddenSize from the Recurrent object? @psyyz10

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Test.scala [hunk 2]
```diff
@@ -59,7 +64,7 @@ object Test {
         index += 1
 
         val validationSet = DataSet.array(labeledInput)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Test.scala]: DataSet.array(labeledInput, sc)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala [hunk 3]
```diff
@@ -16,64 +16,85 @@
...
-        dictionaryLength = dictionaryLength
+
+      val tokensRDD = SequencePreprocess(
+        param.dataFolder + "/train.txt",
+        sc = sc,
+        param.sentFile,
+        param.tokenFile)
+
+      val dictionary = Dictionary(tokensRDD, param.vocabSize)
+      dictionary.save(param.saveFolder)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala]: why save here?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala [hunk 4]
```diff
@@ -16,64 +16,85 @@
...
+
+      val maxTrainLength = tokensRDD.map(x => x.length).collect().max
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala]: use RDD.max()

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala [hunk 5]
```diff
@@ -16,64 +16,85 @@
...
-      val labelPadding = 3999
-      val trainSet = DataSet.array(trainData)
-           .transform(LabeledSentenceToSample(dictionaryLength))
-           .transform(SampleToBatch(
-             batchSize, Some(featurePadding), Some(labelPadding), Some(trainMaxLength)))
-      val validationSet = DataSet.array(valData)
-           .transform(LabeledSentenceToSample(dictionaryLength))
-           .transform(SampleToBatch(
-             batchSize, Some(featurePadding), Some(labelPadding), Some(valMaxLength)))
+      val maxValLength = valtokensRDD.map(x => x.length).collect().max
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala]: use RDD.max()

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala [hunk 6]
```diff
@@ -16,64 +16,85 @@
...
+
+      val traintokens = DataSet.rdd(tokensRDD)
+      val valtokens = DataSet.rdd(valtokensRDD)
+
+      val totalVocabLength = dictionary.getVocabSize() + 1
+      val startIdx = dictionary.getIndex(SentenceToken.start)
+      val endIdx = dictionary.getIndex(SentenceToken.end)
+      val padFeature = Tensor[Float]().resize(totalVocabLength)
+      padFeature.setValue(endIdx + 1, 1.0f)
+      val padLabel = startIdx
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala]: where are padFeature & padLabel used?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Test.scala [hunk 7]
```diff
@@ -38,9 +39,13 @@ object Test {
   def main(args: Array[String]): Unit = {
     testParser.parse(args, new TestParams()).foreach { param =>
 
+      val vocab = Dictionary(param.folder)
+      val conf = Engine.createSparkConf()
+        .setAppName("Test rnn on text")
+        .set("spark.task.maxFailures", "1")
+      val sc = new SparkContext(conf)
       Engine.init
       Engine.setCoreNumber(1)
```
Comment (jason-dai) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Test.scala]: why setCoreNumber(1)?

**Comment by jason-dai** [comment]
This only works on Spark local; please make it clear in readme.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala (hunk 3 above)
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala]: The test file needs to load in saved dictionary.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala (hunk 4 above)
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala]: changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala (hunk 5 above)
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala]: changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala (hunk 6 above)
Comment (JerryYanWan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala]: updated, used it in SampleToBatch.

**Comment by yiheng** [comment]
Please provide example commands for different mode

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/README.md [hunk 8]
```diff
@@ -35,7 +35,7 @@ The input text may look as follows:
 ##Train the Model
 Example command:
 ```bash
-./dist/bin/bigdl.sh -l -- java -cp dist/lib/bigdl-0.1.0-SNAPSHOT-jar-with-dependencies-and-spark.jar com.intel.analytics.bigdl.models.rnn.Train -f /opt/text --nEpochs 30 --learningRate 0.1
+./dist/bin/bigdl.sh -- java -cp bigdl_folder/lib/bigdl-0.1.0-SNAPSHOT-jar-with-dependencies-and-spark.jar com.intel.analytics.bigdl.models.rnn.Train -f /path/inputdata/ -s /path/saveDict --checkpoint /path/model/ --batchSize 12
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/README.md]: why start with java?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/README.md [hunk 9]
```diff
@@ -58,18 +58,16 @@ Each eye that saw him.
 ```
 Example command:
 ```bash
-./dist/bin/bigdl.sh -l -- java -cp dist/lib/bigdl-0.1.0-SNAPSHOT-jar-with-dependencies-and-spark.jar com.intel.analytics.bigdl.models.rnn.Test -f /textdirectory --model /modeldirectory/model.iterationNumber --words 20
+./dist/bin/bigdl.sh -l -- java -cp dist/lib/bigdl-0.1.0-SNAPSHOT-jar-with-dependencies-and-spark.jar com.intel.analytics.bigdl.models.rnn.Test -f /dictDirect --model /modeldirectory/model.iterationNumber --words 20
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/README.md]: why start with java

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/README.md [hunk 10]
```diff
@@ -15,7 +15,7 @@ Please build BigDL referring to [Build Page](https://github.com/intel-analytics/
 ##Prepare the Input Data
 You can download the Tiny Shakespeare Texts corpus from [here](https://raw.githubusercontent.com/karpathy/char-rnn/master/data/tinyshakespeare/input.txt).
 
-After downloading the text, please place it into an appropriate directory (e.g /opt/text/input.txt). The program will later read in the original text file from this directory.
+After downloading the text, please place it into an appropriate directory (e.g /opt/text/input.txt). Please separate it into train.txt and val.txt. In our case, we just select 80 percentage of the input to be train and remaining 20 percentage to be val. The program will later read in the original text file from this directory.
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/README.md]: can you paste the command to do it?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Test.scala [hunk 11]
```diff
@@ -38,56 +40,55 @@ object Test {
...
-          val predictProbDist = logSoftMax.forward(output(output.size(1)))
-            .storage().map(x => math.exp(x).toFloat).toArray
-            .map {
-              var s = 0.0f; d => {
-                s += d; s
-              }
+      val vocabSize = vocab.getVocabSize() + 1
+      val batchSize = param.batchSize
+
+      val rdd = DataSet.array(labeledInput, sc)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Test.scala]: don't use dataset

**Pull request merged**